addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,1204 @@
+    ]
+
+    def _get_max_sequence(self):
+        max_sequence_favourite = self.search([('user_id', '=', self.env.user.id)], order='sequence desc', limit=1)
+        return max_sequence_favourite.sequence + 1 if max_sequence_favourite else -1
+
+
+class ArticleMembers(models.Model):
+    _name = 'knowledge.article.member'
+    _description = 'Article Members'
+    _table = 'knowledge_article_member_rel'
+
+    article_id = fields.Many2one('knowledge.article', 'Article', ondelete='cascade', required=True)
+    partner_id = fields.Many2one('res.partner', index=True, ondelete='cascade', required=True)
+    permission = fields.Selection([
+        ('none', 'No access'),
+        ('read', 'Can read'),
+        ('write', 'Can write'),
+    ], required=True, default='read')
+    # TODO DBE: Make article_permission a related on article_id.inherited_permission
+    article_permission = fields.Selection([
+        ('none', 'No access'),
+        ('read', 'Can read'),
+        ('write', 'Can write'),
+    ], string="Article Permission", compute='_compute_article_permission', store=True,
+        help="This technical field is only used to trigger _check_members constrain")
+    has_higher_permission = fields.Boolean(
+        compute='_compute_has_higher_permission',
+        help="If True, the member has a higher permission then the one set on the article.")
+
+    _sql_constraints = [
+        ('partner_unique', 'unique(article_id, partner_id)', 'You already added this partner in this article.')
+    ]
+
+    def name_get(self):
+        """Override the `name_get` function"""
+        return [(rec.id, "%s" % (rec.partner_id.display_name)) for rec in self]
+
+    @api.constrains('article_permission', 'permission')
+    def _check_members(self):
+        """
+        An article must have at least one member. Since this constrain only triggers if we have at least one member on
+        the article, another validation is done in 'knowledge.article' model.
+        The article_permission related field has been added and stored to force triggering this constrain when
+        article.permission is modified.
+        """
+        parent_members_permission = self.mapped('article_id').mapped('parent_id')._get_article_member_permissions()
+        for member in self:
+            if member.article_id.inherited_permission != 'write':
+                # check on current member article
+                article_write_members = member.article_id.article_member_ids.filtered(
+                        lambda member: member.permission == 'write')
+                if len(article_write_members) > 0:
+                    continue
+                # check on parents members
+                parent_write_members = any(
+                    values['permission'] == 'write' for partner_id, values
+                    in parent_members_permission[member.article_id.parent_id.id].items()
+                    if not member.article_id.is_desynchronized
```

Comment:
is_desynchornized should be added in the sql request. But otherwise, we should still check desynchronized articles. (removing a member of modifying member permissions on desync articles should still trigger a check)